### DISCUSSION TITLE
How to set up Mermaid Rendering Support

### DISCUSSION THREAD

**Discussion by delfireinoso** [open-webui/open-webui#2902] [Q&A] [answered]
I'm unable to find a guide what need to be installed and how to use it. I'm missing it? I would try this new and interesting feature of Open-WebUi

**Comment by silentoplayz** [comment, marked as answer]
You can ask an LLM in Open WebUI to use MermaidJS Diagramming and Charting tool to visualize a thought process or anything you can imagine. Simply ask the LLM to create a Mermaid diagram, such as "Can you create `x` for me using Mermaid?" The LLM will generate the code, and Open WebUI will automatically render any working Mermaid syntax at the end of the response. The possibilities are vast, limited only by what MermaidJS offers, and virtually limitless when paired with an LLM. When paired with the ability to create a structured prompt to guide LLMs towards outputting better MermaidJS syntax (For reference, check out their [documentation](https://mermaid.js.org/intro/)), this can truly be incredibly powerful.

Examples of what to expect: 
![working_example](https://github.com/open-webui/open-webui/assets/50341825/71a2e1cd-70ef-4478-9a1d-8a90c49886dc)
![working_example2](https://github.com/open-webui/open-webui/assets/50341825/4016241e-e69d-46ec-ac03-69c00ba8b311)
![working_example3](https://github.com/open-webui/open-webui/assets/50341825/ed55fc19-3afb-4300-bb1f-b453d9613031)

**Comment by delfireinoso** [comment]
Thanks 

I've made some tests and I get continuously 

> Syntax Error in text
Mermaid version 10.91

Perhaps I need to set up a Prompt to instruct the model to generate a good syntax?

**Reply by silentoplayz** [reply]
That's my best advisory I could give in this case, yes. Some LLMs may struggle, while others may have already been trained on the data contained within MermaidJS docs. Some models will certainly need guided.

**Reply by delfireinoso** [reply]
You have an idea wich models are the best to start?

**Reply by silentoplayz** [reply]
I did forget to mention that the LLM's response should begin with `mermaid` at the start of the Mermaid syntax for it to render correctly. For example, consider this code:

```mermaid
graph TD;
  A-->B;
  B-->C{Decision};
  C--Yes-->D;
  C--No-->E;
  D-->F;
  E-->F;
```

It will generate the following diagram:

```
startAncestor [ start ]
A[A] --> B[B]
B --> C[Decision]
C -->| Yes | D[D]
C -->| No  | E[E]
D --> F[F]
E --> F[F]
```

**Reply by delfireinoso** [reply]
I have success with Codestral

There is a way to save the Image? Or I need to send it to a Mermaid installation?

![image](https://github.com/open-webui/open-webui/assets/28411877/0f1d0d70-38e1-466b-b127-d0c6357519a0)

**Reply by silentoplayz** [reply]
As of now, there's no way to save generated charts, diagrams, etc. outputs via MermaidJS from an LLM in Open WebUI. It's rendered only in the chat.

**Reply by delfireinoso** [reply]
Thanks for your kindly reply.
